File: DSA_Imp/Exercise/MergeKSortedLists.c
#include <iostream>
#include <queue>
#include <vector>
#include <cstdlib>

using namespace std;

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 * 
 * typedef struct ListNode listnode;
 * 
 * listnode* listnode_new(int val) {
 *     listnode* node = (listnode *) malloc(sizeof(listnode));
 *     node->val = val;
 *     node->next = NULL;
 *     return node;
 * }
 */

// Custom comparator to compare nodes based on their value
struct Compare {
    bool operator()(listnode* a, listnode* b) {
        return a->val > b->val;  // Min-heap based on value
    }
};

listnode* mergeKLists(listnode** A, int n1) {
    // Create a min-heap (priority queue)
    priority_queue<listnode*, vector<listnode*>, Compare> minHeap;

    // Push the head of each linked list into the heap
    for (int i = 0; i < n1; i++) {
        if (A[i] != NULL) {
            minHeap.push(A[i]);
        }
    }

    listnode* head = NULL;  // Head of the merged list
    listnode* tail = NULL;  // Tail of the merged list

    // Process the heap
    while (!minHeap.empty()) {
        // Get the smallest node from the heap
        listnode* node = minHeap.top();
        minHeap.pop();

        // Add it to the merged list
        if (head == NULL) {
            head = node;  // First node in the merged list
            tail = head;  // Tail points to head
        } else {
            tail->next = node;
            tail = tail->next;  // Move the tail pointer
        }

        // If the current node has a next node, push it into the heap
        if (node->next != NULL) {
            minHeap.push(node->next);
        }
    }

    return head;  // Return the head of the merged sorted list
}

// Function to print a linked list
void printList(listnode* head) {
    while (head != NULL) {
        cout << head->val << " -> ";
        head = head->next;
    }
    cout << "NULL" << endl;
}

// Helper function to create a new linked list node
listnode* listnode_new(int val) {
    listnode* node = new listnode();
    node->val = val;
    node->next = NULL;
    return node;
}

int main() {
    // Example 1:
    listnode* list1 = listnode_new(1);
    list1->next = listnode_new(10);
    list1->next->next = listnode_new(20);

    listnode* list2 = listnode_new(4);
    list2->next = listnode_new(11);
    list2->next->next = listnode_new(13);

    listnode* list3 = listnode_new(3);
    list3->next = listnode_new(8);
    list3->next->next = listnode_new(9);

    listnode* lists[] = {list1, list2, list3};
    listnode* result = mergeKLists(lists, 3);
    printList(result);

    // Example 2:
    listnode* list4 = listnode_new(10);
    list4->next = listnode_new(12);

    listnode* list5 = listnode_new(13);

    listnode* list6 = listnode_new(5);
    list6->next = listnode_new(6);

    listnode* lists2[] = {list4, list5, list6};
    listnode* result2 = mergeKLists(lists2, 3);
    printList(result2);

    return 0;
}

File: DivideAndConquer/Merge k Sorted Lists.py
# Input: lists = [[1,4,5],[1,3,4],[2,6]]
# Output: [1,1,2,3,4,4,5,6]
# Explanation: The linked-lists are:
# [
#   1->4->5,
#   1->3->4,
#   2->6
# ]
# merging them into one sorted list:
# 1->1->2->3->4->4->5->6

from typing import List, Optional
import heapq

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

    # For debugging, convert linked list to list
    def to_list(self):
        result = []
        node = self
        while node:
            result.append(node.val)
            node = node.next
        return result

class Solution:
    def mergeKLists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:
        if not lists:
            return None

        # Min-Heap to store (value, index, node)
        min_heap = []
        
        # Initialize heap with head nodes of each non-empty list
        for i, node in enumerate(lists):
            if node:
                heapq.heappush(min_heap, (node.val, i, node))

        # Dummy node to help in building the result list
        dummy = ListNode(0)
        current = dummy

        while min_heap:
            # Extract the smallest node
            val, i, node = heapq.heappop(min_heap)
            current.next = node
            current = node  # Move forward in the result list

            # Push the next node from the same list if available
            if node.next:
                heapq.heappush(min_heap, (node.next.val, i, node.next))

        return dummy.next  # Return the merged linked list head